### DISCUSSION TITLE
How do I test my self built SWT bugfix plugin libraries in an Eclipse Php installation ?

### DISCUSSION THREAD

**Discussion by the-snowwhite** [eclipse-platform/eclipse.platform.swt#858] [Q&A] [answered]
In regards to attempting to fix  https://github.com/eclipse-platform/eclipse.platform.swt/issues/851.
I have run into this problem that I have not yet found a way the replace the org.eclipse.swt..... plugins in a standard eclipse
installation with my own homebuilt swt library versions.
So I then can test my current swt fix on a stardard eclipse setup.

When I build with maven via the same command as in the github hook the files are installed to this location on my pc:

    /home/mib/.m2/repository/org/eclipse/swt

and I can see the following files:

    ls org.eclipse.swt/*SNAPSHOT
    maven-metadata-local.xml                            org.eclipse.swt-3.124.200-SNAPSHOT-p2metadata.xml
    org.eclipse.swt-3.124.200-SNAPSHOT.jar              org.eclipse.swt-3.124.200-SNAPSHOT.pom
    org.eclipse.swt-3.124.200-SNAPSHOT-p2artifacts.xml  _remote.repositories

    ls org.eclipse.swt.gtk.linux.x86_64/*SNAPSHOT
    maven-metadata-local.xml
    org.eclipse.swt.gtk.linux.x86_64-3.124.200-SNAPSHOT.jar
    org.eclipse.swt.gtk.linux.x86_64-3.124.200-SNAPSHOT-p2artifacts.xml
    org.eclipse.swt.gtk.linux.x86_64-3.124.200-SNAPSHOT-p2metadata.xml
    org.eclipse.swt.gtk.linux.x86_64-3.124.200-SNAPSHOT.pom
    org.eclipse.swt.gtk.linux.x86_64-3.124.200-SNAPSHOT-sources.jar
    _remote.repositories

Which contains the *.jar files I assume I somehow can replace in an existing eclipse installation.
However I have not yet found any method as how to get this to work.

**Comment by mickaelistria** [comment]
Trying to create a new complete installation with different bundles is not easy. Basically, for something as deep as SWT, it would involve rebuilding the whole IDE.
The simplest and IMO sufficient way to test what you want would be to create a PDE Target Platform that consists of your current installation or target + your newly built SWT artifacts; and then create an "Eclipse Application" Launch configuration and going through its plugin tab to ensure the vanilla SWT bundles are not selected and that yours are selected instead.
HTH

**Reply by the-snowwhite** [reply]
@mickaelistria Thanks for your reply
Funny how you actually answered what I had in my thoughts in the moment I wrote my question but not what I actually
wrote and had intended initially. Pasting a list of files compiled from the master branch probably confused the question also.

What I initially set out to do was to test my bug fix in my standard install of eclipse PHP (updated to 2023-09) on my former development PC that runs the current version of KDE Neon with the standard LTS version of gtk(v3.24.33).
And also test on my newer various other eclipse installations also on my current KDE Neon pc that is manually updated to gtk v3.24.38.
...
And I figured out a solution:
1. checkout the maintenance branch for eclipse 2023-09 (name: R4_29_maintenance).
2. create a new branch fro there and commit my fix.
3. mvn build the binaries and then the sources
4. copy the *.jar files from above mentioned folders under:
    
       /home/user/.m2/repository/org/eclipse/swt

 to a temporary folder.() the file and folder names now have 100 instead of 200 in their names
5. rename the files to:

        org.eclipse.swt_3.124.100.v20230825-1346.jar
        org.eclipse.swt.gtk.linux.x86_64_3.124.100.v20230825-1346.jar
        org.eclipse.swt.gtk.linux.x86_64.source_3.124.100.v20230825-1346.jar

6. then copy them into the plugin folder overwriting the former versions:
on the newer .p2 based installs folder name is:

         /home/user/.p2/pool/plugins

on my elderly installed eclipse php the plugins resided in the eclipse php install folder.
.....
BTW:
If I would go the fallback route of creating a PDE Target Platform I was giving into as a backup plan I would most certainly
need more verbose and detailed Instructions at a lower abstraction level, or else perhaps another week of time scrutinizing
eclipse documentation :-)

**Reply by the-snowwhite** [reply, marked as answer]
An afterthought:
https://github.com/eclipse-platform/eclipse.platform.swt/pull/881#issuecomment-1806604093

**Comment by the-snowwhite** [comment]
A raw more detailed list of the steps:

Prerequisites:

1. Setup an Elipse SWT committer Dev ide. with [eclipse.platform.swt](https://github.com/eclipse-platform/eclipse.platform.swt) and [eclipse.platform.swt.binaries](https://github.com/eclipse-platform/eclipse.platform.swt.binaries) repos.

    (This step is fuzzy at the currently available documentation is somewhat dated.)

    Eclipse RCP_RAP with the extra plugins suggested when cloning [eclipse_platform repo](https://github.com/eclipse-platform/eclipse.platform.swt) via eclipse e-git
    Seemed to do the trick for me this time around with 2023-09_R installer.

2. Setup or have any Standard Eclipse based release you are targeting.
    
    One that you can test and verify the functionality of the patch on,  via a downloaded eclipse installer.
    (In this case Eclipse PHP 2023-09 Linux x86_64) 

3. Setup Maven
    ~$ wget https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.9.3/apache-maven-3.9.3-bin.tar.gz -P /tmp
    sudo tar xf /tmp/apache-maven-*.tar.gz -C /opt
    sudo ln -s /opt/apache-maven-3.9.3 /opt/maven
    sudo nano /etc/profile.d/maven.sh

    check contents:

    mib@neon-am5:~$ cat /etc/profile.d/maven.sh
    export JAVA_HOME=/home/mib/.p2/pool/plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_17.0.8.v20230831-1047/jre
    export M2_HOME=/opt/maven
    export MAVEN_HOME=/opt/maven
    export PATH=${M2_HOME}/bin:${PATH}

Steps 1, 2 and 3 are only need to be done the first time around.

The test loop:

4. Checkout the branch for the release you are targeting in both .swt.binaries and .swt
    and remember to then create a new work-branch from there with any name that describes the purpose.
    For the 2023-09 release the initial branch name was R_29_Maintenance

5. Apply your commit(s) to the work branch. (eventually push your commit to your own fork)

6. Refresh both binaries and swt root in project explorer view. then do project clean from menu
    (that should build some of the project )

7. Optional: test and debug your functionality in one of the provided swt.examples projects.
   If the functionality fails here no need to go further in this list. --> modify patch and debug until success. --> resume from step 4.

8. Cd into the main github folder of the binaries repo. and run

    mvn --batch-mode -Pbuild-individual-bundles -DforceContextQualifier=zzz -Dcompare-version-with-baselines.skip=true -Dmaven.compiler.failOnWarning=true install

9. Cd into the main swt repo folder and run:

    mvn --batch-mode -V -U -DforkCount=1 -Dcompare-version-with-baselines.skip=false -Dmaven.compiler.failOnWarning=true --fail-at-end -DskipNativeTests=false -DfailIfNoTests=false clean install

10.  Extract SWT binaries from build:
    go into /home/mib/.m2/repository/org/eclipse/swt/org.eclipse.swt
    and copy the jar file there into a temp folder.
    go into  /home/mib/.m2/repository/org/eclipse/swt/org.eclipse.swt.gtk.linux.x86_64 folder (or the others in that tree related to your cpu arch).
    and copy the 2 jar files there into the temp folder also.

11. Rename the files to the exact same names as in your eclipse release plugin folder. (in .p2/plugins or eclipseinstallfolder/plugins [if all the plugin files are there]).
    in this process also make a orig folder to act as a backup for the 3 swt-plugin jar's.
    the file names for linux x86_64 2023-09 are:

    mib@neon-am5:~/.p2/pool/test_rel/orig-bak$ ls
    org.eclipse.swt_3.124.100.v20230825-1346.jar
    org.eclipse.swt.gtk.linux.x86_64_3.124.100.v20230825-1346.jar
    org.eclipse.swt.gtk.linux.x86_64.source_3.124.100.v20230825-1346.jar


12. Paste the 3 correctly named jars from the temp folder into the eclipse plugins folder overwriting the ones that came with the distro.

13. Start up your now patched eclipse and do your functional testing of your PR commit(s).

**Comment by the-snowwhite** [comment]
The build automated PR way:

Prerequisites:

1. Setup an Elipse SWT committer Dev ide. with [eclipse.platform.swt](https://github.com/eclipse-platform/eclipse.platform.swt) and [eclipse.platform.swt.binaries](https://github.com/eclipse-platform/eclipse.platform.swt.binaries) repos.

    (This step is fuzzy at the currently available documentation is somewhat dated.)

    Eclipse RCP_RAP with the extra plugins suggested when cloning [eclipse_platform repo](https://github.com/eclipse-platform/eclipse.platform.swt) via eclipse e-git
    Seemed to do the trick for me this time around with 2023-09_R installer.

2. Setup or have any Standard Eclipse based release you are targeting.
    
    One that you can test and verify the functionality of the patch on,  via a downloaded eclipse installer.
    (In this case Eclipse PHP 2023-09 Linux x86_64) 

Creating the draft PR:

3. Checkout the branch for the release you are targeting in both .swt.binaries and .swt
    and remember to then create a new work-branch from there with any name that describes the purpose.
    For the 2023-09 release the initial branch name was R_29_Maintenance

4. Apply your commit(s) to the work branch. then push your commit to your own fork.

5. Optional: Refresh both binaries and swt root in project explorer view. then do project clean from menu
    (that should build some of the project )

6. Optional: test and debug your functionality in one of the provided swt.examples projects.
   If the functionality fails here no need to go further in this list. --> modify patch and debug until success. --> resume from step 4.

7. From github initiate a Draft PR selecting the Branch you are targeting as destination (R_29_Maintenance)

Notice: something like the first commit from #890 is needed for the build to Succeed.

8. Monitor the build and after the build has succeeded-->
Test and external collaborated verification:

9.  download the 2 jar files from the following links to a temp folder on your
    PC.(adjust the PR-883 part to the id of your PR)

    https://ci.eclipse.org/releng/job/eclipse.platform.swt/view/change-requests/job/PR-890/lastSuccessfulBuild/artifact/eclipse.platform.swt.binaries/bundles/org.eclipse.swt.gtk.linux.x86_64/target/

10. Rename the files to the exact same names as in your eclipse release plugin folder. (in .p2/plugins or <eclipseinstallfolder>/plugins [if all the plugin files are there]).
    in this process also make a orig folder to act as a backup for the 2 swt-plugin jar's.
    the file names for linux x86_64 2023-09 are:

    mib@neon-am5:~/.p2/pool/test_rel/orig-bak$ ls
    org.eclipse.swt.gtk.linux.x86_64_3.124.100.v20230825-1346.jar
    org.eclipse.swt.gtk.linux.x86_64.source_3.124.100.v20230825-1346.jar

11. Paste the 3 correctly named jars from the temp folder into the eclipse plugins folder overwriting the ones that came with the distro.

12. Start up your now patched eclipse and do your functional testing of your PR commit(s).
